### PR TITLE
Taboola: Add new adapter

### PR DESCRIPTION
**Pull request by osulzhenko**

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+                validateImp(imp);
+
+                final ExtImpTaboola impExt = parseImpExt(imp);
+                final String type = getBidType(imp).equals(BidType.banner)
+                        ? DISPLAY_ENDPOINT_PREFIX
+                        : BidType.xNative.getName();
+
+                final BidRequest outgoingRequest = createRequest(request, imp, impExt);
+
+                httpRequests.add(HttpRequest.<BidRequest>builder()
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:68]: In GO they create only two `BidRequest`s:
 - one for all imps with `banner`
 - second for all imps with `native`

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        if (imp.getBanner() == null && imp.getXNative() == null) {
+            throw new PreBidException("For Imp ID %s Banner or Native is undefined".formatted(imp.getId()));
+        }
+    }
+
+    private static Imp modifyImp(Imp imp, ExtImpTaboola impExt) {
+        return imp.toBuilder()
+                .tagid(impExt.getPublisherId())
+                .bidfloor(impExt.getBidFloor())
+                .build();
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:129]: In GO they also modify `imp.banner.pos`

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    private BidRequest createRequest(BidRequest request, Imp taboolaImp, ExtImpTaboola impExt) {
+        Imp updatedImp = modifyImp(taboolaImp, impExt);
+        final List<String> blockedAdomain = impExt.getBAdv();
+        final List<String> blockedAdvCat = impExt.getBCat();
+        Site newSite;
+        if (request.getSite() == null) {
+            newSite = Site.builder()
+                    .id(impExt.getPublisherId())
+                    .name(impExt.getPublisherId())
+                    .domain(StringUtils.defaultString(impExt.getPublisherDomain()))
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:91]: In Go they use `request.site.domain` here also.

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                    .publisher(Publisher.builder().id(impExt.getPublisherId()).build())
+                    .build();
+        } else {
+            newSite = request.getSite();
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:95]: You need to modify existing site.

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:96]: Can be simplified:
```
Optional.ofNullable(request.getSite())
      .map(Site::toBuilder)
      .orElseGet(Site::builder)
      .id(...)
      ...
      .build()
```

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    private List<BidderBid> bidsFromResponse(BidRequest bidRequest,
+                                             BidResponse bidResponse) {
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .map(bid -> BidderBid.of(bid, getBidType(bidRequest.getImp().get(0)), bidResponse.getCur()))
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:166]: In GO they used all of imps to recognize `BidType`.

Comment (CTMBNara) [reply src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:166]: Also, if there is some problems, they just skip that bid. Do not skip all of bids on first error.

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/taboola/ExtImpTaboola.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    String publisherDomain;
+    @JsonProperty("tagid")
+    String tagId;
+    @JsonProperty("bidfloor")
+    BigDecimal bidFloor;
+    @JsonProperty("bcat")
+    List<String> bCat;
+    @JsonProperty("badv")
+    List<String> bAdv;
+}
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/taboola/ExtImpTaboola.java:24]: Missing `pageType` and `position`

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+                final BidRequest outgoingRequest = createRequest(request, imp, impExt);
+
+                httpRequests.add(HttpRequest.<BidRequest>builder()
+                        .method(HttpMethod.POST)
+                        .uri(buildEndpointUrl(impExt, type, outgoingRequest.getSite().getDomain()))
+                        .body(mapper.encodeToBytes(outgoingRequest))
+                        .headers(headers)
+                        .payload(outgoingRequest)
+                        .build());
```
[the remark this answers is not in the document]
Comment (osulzhenko) [reply src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java]: I have added all of this as a `try-catch` block due to avoid duplicate requests. Because if we get an error with an invalid imp, we can still get BidRequest after this. So, in my opinion, it should all be in `try-catch`.

File: src/main/resources/static/bidder-params/taboola.json [hunk 9]
```diff
@@ -0,0 +1,40 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "title": "Taboola Adapter Params",
+  "description": "A schema which validates params accepted by the Taboola adapter",
+
```
Comment (CTMBNara) [inline src/main/resources/static/bidder-params/taboola.json:5]: Please, remove this empty line. `cmd+option+L` or `ctrl+alt+L`, if you work in idea.

File: src/main/resources/bidder-config/taboola.yaml [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+      supported-vendors:
+      vendor-id: 42
+    userSync:
+      cookie-family-name: taboola
+      redirect:
+        url: https://trc.taboola.com/sg/ps/1/cm?gdpr={{.GDPR}}&gdpr_consent={{.GDPRConsent}}&us_privacy={{.USPrivacy}}&redirect={{.RedirectURL}}
+        support-cors: false
+        uid-macro: '{TUID}'
+      iframe:
+        url: https://cdn.taboola.com/scripts/ps-sync.html?gdpr={{.GDPR}}&gdpr_consent={{.GDPRConsent}}&us_privacy={{.USPrivacy}}&redirect={{.RedirectURL}}
```
Comment (CTMBNara) [inline src/main/resources/bidder-config/taboola.yaml:22]: Please, replace GO macros with ours
https://github.com/prebid/prebid-server-java/blob/7a5b4c36b29e169b4610a22dc538bc6e3fac79f0/src/main/java/org/prebid/server/bidder/UsersyncUtil.java#L11

File: src/main/resources/bidder-config/taboola.yaml [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+        support-cors: false
+        uid-macro: '{TUID}'
```
Comment (CTMBNara) [inline src/main/resources/bidder-config/taboola.yaml:24]: `'<TUID>'`

File: src/main/java/org/prebid/server/proto/openrtb/ext/request/taboola/ExtImpTaboola.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    @JsonProperty("tagid")
+    String tagId;
+    @JsonProperty("bidfloor")
+    BigDecimal bidFloor;
+    @JsonProperty("bcat")
+    List<String> bCat;
+    @JsonProperty("badv")
+    List<String> bAdv;
+    String pageType;
+    Integer position;
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/proto/openrtb/ext/request/taboola/ExtImpTaboola.java:25]: Please, add empty lines between fields.

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (CTMBNara) [reply src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java]: Just use `continue` keyword in this case.

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java (hunk 1 above)
Comment (CTMBNara) [reply src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java]: Please pay attention to this ^^

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        return bidsFromResponse(bidRequest, bidResponse);
+    }
+
+    private List<BidderBid> bidsFromResponse(BidRequest bidRequest,
+                                             BidResponse bidResponse) {
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .flatMap(seatBid -> seatBid.getBid().stream()
+                        .flatMap(bid -> bidRequest.getImp().stream()
+                                .map(imp -> BidderBid.of(bid, getBidType(imp), bidResponse.getCur()))))
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:177]: Please, return previous solution.

Also, you can't use same `getBidType` method for response.
Pay attention that when you will throw exception about something wrong in the current bid, you need to catch it, add an error to errors list and continue process other bids.

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+            if (imp.getId().equals(impId)) {
+                if (imp.getBanner() != null) {
+                    return BidType.banner;
+                } else if (imp.getXNative() != null) {
+                    return BidType.xNative;
+                }
+                return BidType.banner;
+            }
+        }
+        throw new PreBidException("Failed to find impression \"%s\"".formatted(impId));
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:222]: `Failed to find banner/native impression "%s" `

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .map(bid -> resolveBidderBid(bidResponse.getCur(), bidRequest.getImp(), bid, errors))
+                .toList();
+    }
+
+    private BidderBid resolveBidderBid(String currency, List<Imp> imps, Bid bid, List<BidderError> errors) {
+        try {
+            BidType bidType = resolveBidType(bid.getImpid(), imps);
+            return BidderBid.of(bid, bidType, currency);
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:204]: Please, inline `resolveBidType` call.

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+                                             BidResponse bidResponse,
+                                             List<BidderError> errors) {
+
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .map(bid -> resolveBidderBid(bidResponse.getCur(), bidRequest.getImp(), bid, errors))
+                .toList();
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:198]: Add `.filter(Objects::nonNull)` before collecting bids.

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+        } catch (DecodeException e) {
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private List<BidderBid> extractBids(BidRequest bidRequest, BidResponse bidResponse) {
+        if (bidResponse == null || CollectionUtils.isEmpty(bidResponse.getSeatbid())) {
+            return Collections.emptyList();
+        }
+        final List<BidderError> errors = new ArrayList<>();
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:184]: You need to extract this lists to `makeBids` method.

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java [hunk 18]
```diff
@@ -0,0 +1,326 @@
...
+
+        final BidRequest bidRequest = BidRequest
+                .builder().imp(List.of(impBanner)).site(null).build();
+
+        final Site expectedSite = Site.builder()
+                .id("token")
+                .name("token")
+                .domain("test.com")
+                .publisher(Publisher.builder().id("token").build())
+                .build();
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:107]: Pls, move expected results to `then` block.

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java [hunk 19]
```diff
@@ -0,0 +1,326 @@
...
+        assertThatIllegalArgumentException().isThrownBy(() -> new TaboolaBidder("incorrect.endpoint",
+                "http://localhost:8090",
+                jacksonMapper));
+    }
+
+    @Test
+    public void makeHttpRequestsShouldTakeCurrencyAmountFromImp() {
+        // given
+        final BidRequest bidRequest = givenBidRequest(
+                impCustomizer -> impCustomizer.banner(Banner.builder().build()).bidfloorcur("USD"));
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:68]: No need to set and then assert `bidfloorcur`. Adapter doesn't modify it.
Same for other occurrences.

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java [hunk 20]
```diff
@@ -0,0 +1,326 @@
...
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).hasSize(1)
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+                .flatExtracting(BidRequest::getImp)
+                .extracting(Imp::getBidfloor, Imp::getBidfloorcur)
+                .containsExactly(tuple(BigDecimal.TEN, "USD"));
+    }
+
+    @Test
+    public void makeHttpRequestWithEmptySiteShouldTakeDataForSiteFromImpValues() {
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:83]: What about case when `Site` present in initial request?

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+        final Map<MediaType, List<Imp>> mediaTypeToImps = new HashMap<>();
+        ExtImpTaboola extImpTaboola = null;
+
+        for (Imp imp : request.getImp()) {
+            try {
+                extImpTaboola = parseImpExt(imp);
+            } catch (PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+                continue;
+            }
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:75]: Cover with unit tests.

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 22]
```diff
@@ -0,0 +1,271 @@
...
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage());
+        }
+    }
+
+    private static Imp modifyImp(Imp imp, ExtImpTaboola impExt) {
+        final String impExtTagId = impExt.getTagId();
+        final UpdateResult<String> resolvedTagId = StringUtils.isNotEmpty(impExtTagId)
+                ? UpdateResult.updated(impExtTagId)
+                : UpdateResult.unaltered(imp.getTagid());
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:122]: Cover with unit tests.

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 23]
```diff
@@ -0,0 +1,271 @@
...
+                .orElseGet(Site::builder)
+                .id(impExtPublisherId)
+                .name(impExtPublisherId)
+                .domain(resolveDomain(impExt.getPublisherDomain(), request))
+                .publisher(Publisher.builder().id(impExtPublisherId).build())
+                .build();
+
+        final ExtRequest extRequest = StringUtils.isNotEmpty(impExtPageType)
+                ? createExtRequest(impExtPageType)
+                : request.getExt();
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:163]: Cover with unit tests.

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java [hunk 24]
```diff
@@ -0,0 +1,326 @@
...
+        Result<List<HttpRequest<BidRequest>>> result = taboolaBidder.makeHttpRequests(bidRequest);
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue())
+                .extracting(HttpRequest::getHeaders)
+                .flatExtracting(MultiMap::entries)
+                .extracting(Map.Entry::getKey, Map.Entry::getValue)
+                .containsExactlyInAnyOrder(
+                        tuple(HttpUtil.CONTENT_TYPE_HEADER.toString(), HttpUtil.APPLICATION_JSON_CONTENT_TYPE),
+                        tuple(HttpUtil.ACCEPT_HEADER.toString(), HttpHeaderValues.APPLICATION_JSON.toString())
+                );
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:157]: No need to assert headers. Adapter doesn't modify it.

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java [hunk 25]
```diff
@@ -0,0 +1,326 @@
...
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).hasSize(2)
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+                .flatExtracting(BidRequest::getImp)
+                .extracting(Imp::getBidfloor, Imp::getBidfloorcur)
+                .containsExactly(tuple(BigDecimal.TEN, "EUR"), tuple(BigDecimal.TEN, "USD"));
+    }
+
+    @Test
+    public void makeHttpRequestShouldContainProperUriAndHeader() {
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:142]: Cover case with `native`.

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java [hunk 26]
```diff
@@ -0,0 +1,326 @@
...
+                .flatExtracting(MultiMap::entries)
+                .extracting(Map.Entry::getKey, Map.Entry::getValue)
+                .containsExactlyInAnyOrder(
+                        tuple(HttpUtil.CONTENT_TYPE_HEADER.toString(), HttpUtil.APPLICATION_JSON_CONTENT_TYPE),
+                        tuple(HttpUtil.ACCEPT_HEADER.toString(), HttpHeaderValues.APPLICATION_JSON.toString())
+                );
+        assertThat(result.getValue())
+                .extracting(HttpRequest::getUri)
+                .containsExactly("https://display.bidder.taboola.com/OpenRTB/PS/auction/localhost-test.com/token")
+                .doesNotContain(BidType.banner.getName());
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:161]: Redundant `doesNotContain`.

File: src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java [hunk 27]
```diff
@@ -0,0 +1,271 @@
...
+            case NATIVE -> NATIVE_ENDPOINT_PREFIX;
+
+            // should never happen
+            default -> throw new AssertionError();
+        };
+
+        return endpointTemplate
+                .replace("{{Host}}", domain)
+                .replace("{{MediaType}}", type)
+                .replace("{{PublisherID}}", StringUtils.defaultString(publisherId, ""));
```
Comment (CTMBNara) [inline src/main/java/org/prebid/server/bidder/taboola/TaboolaBidder.java:210]: Cover with unit test case with empty string.

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java [hunk 28]
```diff
@@ -0,0 +1,326 @@
...
+            UnaryOperator<Imp.ImpBuilder> impCustomizer) {
+
+        return bidRequestCustomizer.apply(BidRequest.builder()
+                        .imp(List.of(givenImp(impCustomizer))))
+                .build();
+    }
+
+    private static BidRequest givenBidRequest(UnaryOperator<Imp.ImpBuilder> impCustomizer) {
+        return givenBidRequest(identity(), impCustomizer);
+    }
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:317]: What about to use this methods? It will allow you to easily use multiple imps in one test case.
```
    private static BidRequest givenBidRequest(UnaryOperator<BidRequest.BidRequestBuilder> bidRequestCustomizer,
                                              Imp... imps) {

        return bidRequestCustomizer.apply(BidRequest.builder().imp(List.of(imps))).build();
    }

    private static Imp givenImp(UnaryOperator<Imp.ImpBuilder> impCustomizer,
                                UnaryOperator<ExtImpTaboola.ExtImpTaboolaBuilder> extImpCustomizer) {

        return impCustomizer
                .apply(Imp.builder().ext(mapper.valueToTree(extImpCustomizer.apply(ExtImpTaboola.builder()).build())))
                .build();
    }

```

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:1]: Please try to put the tests in the order they "appear" in the adapter code.

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java [hunk 29]
```diff
@@ -0,0 +1,326 @@
...
+        // when
+        final Result<List<BidderBid>> result = taboolaBidder.makeBids(httpCall, bidRequest);
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).isEmpty();
+    }
+
+    @Test
+    public void makeBidsShouldReturnValidBidResponseWithNativeWhenSingleImp() throws JsonProcessingException {
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:207]: I don't think that `bidType` can depends on imps count.
Same for other tests.

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java [hunk 30]
```diff
@@ -0,0 +1,326 @@
...
+        // when
+        final Result<List<BidderBid>> result = taboolaBidder.makeBids(httpCall, bidRequest);
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue())
+                .extracting(BidderBid::getBid)
+                .containsOnly(Bid.builder()
+                        .impid("123")
+                        .build());
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:225]: No need to assert `bid`. Same for other tests.

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java [hunk 31]
```diff
@@ -0,0 +1,326 @@
...
+        assertThat(result.getValue())
+                .extracting(BidderBid::getType)
+                .containsExactly(BidType.xNative);
+    }
+
+    @Test
+    public void makeBidsShouldReturnValidBidResponseWithBannerWhenRequestHaveTwoImp() throws JsonProcessingException {
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:232]: Multiple imps can't have same `imp.id`

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java [hunk 32]
```diff
@@ -0,0 +1,400 @@
...
+        Result<List<HttpRequest<BidRequest>>> result = taboolaBidder.makeHttpRequests(bidRequest);
+
+        // then
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).hasSize(2)
+                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
+                .flatExtracting(BidRequest::getImp)
+                .extracting(Imp::getBidfloor)
+                .containsExactlyInAnyOrder(BigDecimal.ONE, BigDecimal.TEN);
+    }
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:252]: What about this approach?
```
    @Test
    public void makeHttpRequestShouldSplitInTwoSeparateRequests() {
        // given
        final BidRequest bidRequest = givenBidRequest(
                givenBannerImp(identity(), identity()),
                givenImp(imp -> imp.xNative(Native.builder().build()), identity()),
                givenBannerImp(identity(), identity()),
                givenImp(imp -> imp.xNative(Native.builder().build()), identity()));

        // when
        Result<List<HttpRequest<BidRequest>>> result = taboolaBidder.makeHttpRequests(bidRequest);

        // then
        assertThat(result.getErrors()).isEmpty();
        assertThat(result.getValue()).hasSize(2)
                .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
                .extracting(BidRequest::getImp)
                .anySatisfy(imps -> assertThat(imps)
                        .hasSize(2)
                        .allSatisfy(imp -> assertThat(imp.getBanner()).isNotNull()))
                .anySatisfy(imps -> assertThat(imps)
                        .hasSize(2)
                        .allSatisfy(imp -> assertThat(imp.getXNative()).isNotNull()));
    }
```

File: src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java [hunk 33]
```diff
@@ -0,0 +1,400 @@
...
+                        .bidFloor(BigDecimal.TEN)
+                        .publisherDomain("test.com")
+                        .tagId("tag_video"));
+        final Imp impNative = givenImp(impCustomizer ->
+                        impCustomizer.id("124").xNative(Native.builder().build()),
+                extImpCustomizer -> extImpCustomizer
+                        .publisherId("publisher_native_Id")
+                        .bidFloor(BigDecimal.TEN)
+                        .publisherDomain("test.com")
+                        .tagId("tag_native"));
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/taboola/TaboolaBidderTest.java:346]: This imp (id=124) is redundant. It does zero affect on test (none of the bids corresponds to this imp).
Please, remove it.

**Review by CTMBNara — CHANGES_REQUESTED** [review]
